src/ipc/ipc4/handler.c

```diff
@@ -65,7 +65,7 @@ struct ipc4_msg_data {
 static struct ipc4_msg_data msg_data;
```

Comment:
This is what's happening:
```
IPC4:

core 0					core 1

IPC
ipc_task_ops.run()
ipc_do_cmd()
ipc_platform_do_cmd()
ipc_cmd()
mask |= INLINE
mask |= SECONDARY
IDC
					idc_ipc()
					ipc_cmd()
					ipc_msg_send()
					list_item_append()
					mask &= ~SECONDARY
if (!(mask & SECONDARY))
	ipc_msg_send()
	list_item_append()

ipc_task_ops.complete()
ipc_complete_task()
```
and it isn't happening on IPC3 because there we always check return value on the primary core and it's set to 1 when delegating to a secondary core. In principle I don't think you need to take the spinlock below, because in `idc_handler()` first the message is added to the list and only then the `IPC_TASK_SECONDARY_CORE` flag is cleared, so at least one of the indicators is always there, but it actually does "seem to add correctness" using a spinlock here too, so, I wouldn't object it.
The only request here - I think we need a comment to explain, that just checking the flag isn't enough, because the secondary core could've outrun us and already cleared the flag, so we need to check the list too.
But - can it not happen that the IPC worker also preempts us, sends the message out and removes it from the list, so we re-add it again? Maybe we need a more robust flag like that return code in IPC3.

Reply:
@lyakh  In term of lock, I first didn't add it but added it later for some reviewer may advise to add it.  And the issue that IPC worker also preempts ipc thread, I first thought the IPC thread has higher priority than worker thread, so it is no problem. But later I still got the ipc time out issue in long time test. I added some log and found it was caused by your predict. Does zephyr schedule thread based not only on priority ? Consider starve thread ?  Thanks!

Reply:
Sorry, I shouldn't have used "preempt" - I meant when running on 2 cores if the worker could run on core 1 before core 0 checks the mask as in pseudo-code above.
And BTW, it just occurred to me, that our IPC sending work can run on any core and that is probably not what we want?